bpo-27645: Supporting native backup facility of SQLite

lelit opened the pull request:
This is a set of patches (rebased on current master) that adds a new `backup()` method on `sqlite3.Connection`.

<!-- issue-number: bpo-27645 -->
https://bugs.python.org/issue27645
<!-- /issue-number -->

lelit committed:
Expose the SQLite Online Backup API as Connection.backup()

See issue #27645.

lelit committed:
Preliminary documentation for sqlite3.Connection.backup()

lelit committed:
Reduce code nesting depth by using a finally label, as other functions do

lelit committed:
Wrap SQLite calls between BEGIN_ALLOW_THREADS and END_ALLOW_THREADS

Suggested by Aviv Palivoda.

lelit committed:
Add an entry in Misc/NEWS related to issue #27645

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username " requested by the form is the "Login name" field under "Your Details".
4. If you just signed the CLA, please **wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

lelit commented:
I configured my GitHub username on b.p.o.

berkerpeksag commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,34 @@ Connection Objects
                  f.write('%s\n' % line)
 
 
+   .. method:: backup(filename[, pages, progress])
```
I'd like to make pages and progress keyword-only.

lelit replied:
Ok.

berkerpeksag commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,34 @@ Connection Objects
 
+   .. method:: backup(filename[, pages, progress])
+
+      This method exposes the `API`__ that allows to make a backup of a SQLite
```
No need to add a link to SQLite docs.

lelit replied:
Ok.

berkerpeksag commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,34 @@ Connection Objects
+
+         con = sqlite3.connect('existing_db.db')
+         con.backup('copy_of_existing_db.db', 1, progress)
+
```
Needs:

```rst
.. versionadded:: 3.7
```

lelit replied:
Ok.

berkerpeksag commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,34 @@ Connection Objects
+      database into the mandatory argument *filename*, even while it's being accessed
+      by other clients, or concurrently by the same connection.
+
+      By default, or when *pages* is either ``0`` or a negative integer, the entire
```
I wonder if we really need to expose this. The default value should cover most of the use cases.

lelit replied:
Do you mean that I should remove the pages argument? But if I do that, also the progress one becomes pointless, because with the default value there will be at most one callback, when the engine completes the backup. What am I missing?

berkerpeksag replied:
I think it's fine to keep *pages* for huge databases.

berkerpeksag commented on Misc/NEWS:
```diff
@@ -34,6 +36,9 @@ Core and Builtins
 
 - bpo-29546: Improve from-import error message with location
 
+- Issue #7063: Remove dead code from the "array" module's slice handling.
```
This needs to be deleted.

berkerpeksag commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,6 +1478,91 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+    rc = sqlite3_open(filename, &bckconn);
```
https://github.com/python/cpython/pull/359 could be useful here.

lelit replied:
Not sure what you are suggesting: should I take the same approach of conditionally using sqlite3_open_v2() instead?

berkerpeksag replied:
I was trying to say that this should be merged *after* #359.

palaviv replied:
This don't actually have any effect. `sqlite3_open(filename, db)` is the same as `sqlite3_open_v2(filename, db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, NULL)`

berkerpeksag commented on Misc/NEWS:
```diff
@@ -10,6 +10,8 @@ What's New in Python 3.7.0 alpha 1?
 -----------------
 
+- bpo-27645: sqlite3.Connection now exposes a backup() method.
+
```
Add "Patch by Your Name." to the next line.

lelit replied:
Ok.

berkerpeksag commented on Lib/test/test_sqlite.py:
```diff
@@ -18,7 +18,7 @@ def load_tests(*args):
                                factory.suite(), transactions.suite(),
                                hooks.suite(), regression.suite(),
-                               dump.suite()])
+                               dump.suite(), backup.suite()])
```
Nitpick: Add this to its own line so we won't clutter `git blame` next time we add something.

lelit replied:
Ok.

berkerpeksag commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,34 @@ Connection Objects
+         import sqlite3
+
+         def progress(remaining, total):
+             print("Copied %d of %d pages..." % (total-remaining, total))
```
We can use f-strings here.

lelit replied:
Ok.

lelit committed:
Remove extraneous entry introduced in d4439d6c5b2a67682f3db74d395c603f7861361d

lelit committed:
Attribute the patch to myself

lelit committed:
Make pages and progress keyword-only args in sqlite3.Connection.backup()

lelit committed:
Remove reference to SQLite docs in Connection.backup() documentation

lelit committed:
Use f-string in the sqlite3.Connection.backup() example

lelit committed:
Explicitly mark sqlite3.Connection.backup() as added in v3.7

lelit committed:
Reduce chances of cluttering future “git blame”

lelit commented:
Thank you @berkerpeksag!

palaviv commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,6 +1478,92 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+            /* Sleep for 250ms if there are still further pages to copy */
+            if (rc == SQLITE_OK || rc == SQLITE_BUSY || rc == SQLITE_LOCKED) {
+                Py_BEGIN_ALLOW_THREADS
+                sqlite3_sleep(250);
```
Are we sure we need to sleep when `SQLITE_OK` is returned? I know this is what is done in the [example](https://www.sqlite.org/backup.html) but we do allow other threads to take the lock between calls to `sqlite3_backup_step`.
In addition we might want to have some kind of timeout here. We can block forever if someone holds the db lock.

lelit replied:
Yes, I think we should: the step function returns OK when it “successfully copies N pages and there are still more pages to be copied”.
Not sure about the timeout: maybe we could check whether there is no progress in, say, X calls to the step function?

lelit replied:
Reconsidering, I think you're right here: there's no reason to delay next iteration when the result was `SQLITE_OK`.

palaviv commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,6 +1478,92 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+            rc = sqlite3_backup_step(bckhandle, pages);
+            Py_END_ALLOW_THREADS
+
+            if (progress != Py_None) {
```
We should call `progress` only when `sqlite3_backup_step` return `SQLITE_OK`

lelit replied:
Not sure about that: maybe a better option would be pass the result status to the `progress` callback, so that it can eventually signal the error and do some cleanup (think about closing a progress dialog).

lelit replied:
Now the progress callback receives three arguments.

palaviv commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,6 +1478,92 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+        Py_INCREF(Py_None);
+        retval = Py_None;
+    } else {
+        /* TODO: should the (probably incomplete/invalid) backup be removed here? */
```
I think you should delete the backup

lelit replied:
Ok, will do that.

Thanks a lot for the review!

lelit replied:
I need some help here: I expected to find an abstract/portable `PyOS_UnlinkFile()`, but couldn't find it.
What is the recommended idiom to remove a file from C? Is the only way to "import" the `os` module and then `PyObject_CallFunction()` on its `unlink` member?

lelit replied:
I took the simplest way for now, using unlink(2) from `unistd.h`. Any hint for a better way is welcome.

palaviv commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,6 +1478,92 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+                                           sqlite3_backup_pagecount(bckhandle))) {
+                    /* User's callback raised an error: interrupt the loop and
+                       propagate it. */
+                    rc = -1;
```
You should add a test for the case that `progress` raise an error

lelit replied:
Ok, will do!

lelit replied:
Done.

palaviv commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,6 +1478,92 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+    }
+
+    if (rc != -1) {
+        rc = _pysqlite_seterror(bckconn, NULL);
```
I am not sure about this but from the documentation `sqlite3_backup_step`  will not set an error code on error but will return one. I can see that in the example they do the same as here but I am not sure if this is the right behavior. Would we receive `SQLITE_NOMEM` error from `sqlite3_backup_step`?

lelit replied:
Will try to understand the point.

lelit replied:
I tried to address this concern in my latest patches.

palaviv commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,6 +1478,92 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+    bckhandle = sqlite3_backup_init(bckconn, "main", self->db, "main");
```
Shouldn't database name be a parameter?

lelit replied:
That's not really the _database name_: accordingly to the [documentation](http://sqlite.org/c3ref/backup_finish.html#sqlite3backupinit) that name denotes the target, that can either be "main" for the main storage, "temp" for the temporary storage or the name of the _attached_ database.
Of course we can easily expose that as a (default) argument, but I did not see as necessary when I coded the feature.

lelit replied:
I added a `source_name` and a `dest_name` parameters.

lelit replied:
Uhm, further investigation seems to reveal that the `dest_name` is pointless: _attached_ databases are **not** a persistent property of a database, so that we cannot allow the following:
```python
bck = sqlite.connect('/tmp/thebackup')
bck.execute(f"ATTACH DATABASE '/tmp/theattachment' AS attached_db")
bck.commit()

db = sqlite.connect('/tmp/database')
db.backup('/tmp/thebackup', dest_name='attached_db')
```

Changed back to a single `name` argument to specify the _source database_...

lelit committed:
Parametrize the name of the database involved in the backup

Suggested by Aviv Palivoda.

lelit committed:
Test propagation of exception raised in backup's progress callback

lelit committed:
Use a better name for variable

lelit committed:
Do not delay next iteration if the result was OK

When sqlite3_backup_step() returns OK it means that it was able to do it's work but there are
further pages to be copied: in such case it's better to immediately start the next iteration,
without opening a window that allows other threads to go on and possibly locking the database.

Suggested by Aviv Palivoda.

berkerpeksag commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,6 +1478,94 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
 }
 
 static PyObject *
+pysqlite_connection_backup(pysqlite_Connection* self, PyObject* args, PyObject* kwds)
```
If I read [SQLite release notes](https://www.sqlite.org/changes.html) correctly, `sqlite3_backup_init` has been added in 3.6.11 so we need to expose this only if SQLite version is 3.6.11 or newer.

lelit replied:
Ok, is it better to omit the whole function or rather raise an `NotImplementedError`?

berkerpeksag replied:
Let's omit the whole function. This is what we do for other version specific features in the stdlib. Please also update the documentation to note that this feature is available from SQLite 3.6.11.

lelit replied:
Ok, done. Thank you!

berkerpeksag commented on Lib/sqlite3/test/backup.py:
```diff
@@ -0,0 +1,127 @@
+from tempfile import NamedTemporaryFile
+import unittest
+
+class BackupTests(unittest.TestCase):
```
There should be a `@unittest.skipIf(...)` decorator to skip the whole test case if SQLite version is not 3.6.11 or newer.

berkerpeksag replied:
Please add tests to make sure that keyword-only parameters are really keyword-only:

```py
self.cx.backup(bckfn.name, 1, lambda x, y: None)  # should raise TypeError
```

lelit committed:
Omit the backup method when underlying SQLite library is older than 3.6.11

lelit committed:
Assert that the non-mandatory arguments are keyword-only

palaviv commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,42 @@ Connection Objects
+      database is copied in a single step; otherwise the method performs a loop
+      copying up to the specified *pages* at a time.
+
+      If *progress* is specified, it must either ``None`` or a callable object that
```
either be \``None``

lelit committed:
Pass also the current status of the ongoing backup to the progress callback

Add new SQLITE_DONE integer constant to the module, that could be used by the callback to
determine whether the backup is terminated.

lelit committed:
Slightly different way handling backup step's error state

This should address Aviv's concerns about how the sqlite3_backup_step() errors are handled.

lelit committed:
When an error occurs while the backup is going on, remove the target file

lelit commented:
Is there anything else I could/should do to see this accepted?

dgilman commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,43 @@ Connection Objects
+
+      This method makes a backup of a SQLite database into the mandatory argument
+      *filename*, even while it's being accessed by other clients, or concurrently by
+      the same connection.
```
Maybe mention that backups work for on-disk and in-memory databases like the sqlite3 C api docs mention?

lelit commented:
FYI, I rebased these commits on top of current master: https://github.com/lelit/cpython/tree/sqlite-backup-api-v3

lelit commented:
@Mariatta I'm confused by that "added `needs rebase` and removed `needs rebase`, what should I do?
Also, given that current implementation (rebased on master of three weeks ago) lives in a different branch (https://github.com/lelit/cpython/compare/sqlite-backup-api-v3) should I do something wrt this PR?

Mariatta commented:
@lelit We can only review PRs in CPython repo. Currently this PR has conflict. If the conflict is resolved in a different branch, maybe you can create the PR for the other branch?
Although I would prefer that the conflict is resolved in the same branch so we get to keep the past discussions in the same PR.

lelit commented:
Does that mean you'd prefer me to rebase the branch I opened the PR with?

brettcannon commented:
@lelit There's a conflict because you edited Misc/NEWS and it will no longer merge cleanly as-is. See https://devguide.python.org/committing/?highlight=news.d#what-s-new-and-news-entries on how to fix this.

lelit commented:
@brettcannon, I rebased the branch into https://github.com/lelit/cpython/compare/sqlite-backup-api-v3
rewriting the NEWS change using the new tool.

As said in [this comment](https://github.com/python/cpython/pull/377#issuecomment-334573880) I'm sorry about the fuss, but it's not clear to me whether I should have closed this PR and reopened a new one on the v3 branch, or if instead I should have rewritten the same v2 branch.

As always, I'm willing to do whatever is easier/more correct for you!

brettcannon commented:
@lelit I'm not sure what 'v2' and 'v3' are referring to (some personal branches in your fork you have?), but if you look at https://github.com/python/cpython/pull/377/files you will see that the branch your PR is based on has not been updated in regards to the NEWS file.

If it's easier, you can just close this PR, open a new one if you have some other branch that's in better shape, and then refer to this old PR in the new PR's opening comment.

lelit commented:
Ok, doing that immediately.

lelit commented:
As suggested by Brett, I'm closing this PR, replaced by #4238.
